### PR TITLE
Show usage of IDOManager::EnumDownloads

### PR DESCRIPTION
**Pull request by shishirb-MSFT**

### CODE REVIEW

File: Program.cs [hunk 1]
```diff
@@ -148,5 +123,106 @@ private static DODownloadRanges ParseDownloadRanges(string arg)
                 return new DODownloadRanges(offsetLengths);
             }
         }
+
+        private static int ExecDownload(Options options)
```
Comment (shishirb-MSFT) [inline Program.cs:127]: >ExecDownload [](http://example.com/codeflow?start=27&length=12)

This is a straight code move from Main(). #Resolved

File: DODownloadFactory.cs [hunk 2]
```diff
@@ -64,5 +79,56 @@ private void SetCommonProperties(IDODownload downloadObj, DOFile file)
...
+        private List<IDODownload> GetDownloads(DO_DOWNLOAD_ENUM_CATEGORY? enumCategory)
+        {
+            var downloads = new List<IDODownload>();
+            IDOManager manager = GetDOManager();
+            var category = IntPtr.Zero;
+            try
+            {
+                if (enumCategory.HasValue)
+                {
+                    category = Marshal.AllocCoTaskMem(Marshal.SizeOf<DO_DOWNLOAD_ENUM_CATEGORY>());
```
Comment (JeffreySaathoff) [inline DODownloadFactory.cs:92]: >AllocCoTaskMem [](http://example.com/codeflow?start=39&length=14)

Is AllocCoTaskMem really required? It's an in-param with no transfer of buffer ownership. I would rather pass enumCategory.Value directly to manager.EnumDownloads if we can.

What are we using for interop for this parameter? #Closed

Comment (shishirb-MSFT) [reply DODownloadFactory.cs:92]: Public docs says that declaring this param as a `ref <struct> <param-name>` and passing `ref <variable>` at the call site should work, but it didn't. DO client receives an all-zero struct in this case.

**Review by JeffreySaathoff — APPROVED** [review]
:shipit:

**Pull request merged**